server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/JamesMailSpooler.java

```diff
@@ -101,47 +92,38 @@ public void configure(HierarchicalConfiguration<ImmutableNode> config) {
     @PostConstruct
     public void init() {
         LOGGER.info("init...");
-        queue = queueFactory.createQueue(MailQueueFactory.SPOOL);
-        spooler = Schedulers.fromExecutor(Executors.newFixedThreadPool(numThreads, NamedThreadFactory.withName("spooler")));
-        LOGGER.info("uses {} Thread(s)", numThreads);
-        run();
+        LOGGER.info("Concurrency level is {}", concurrencyLevel);
+        queue = queueFactory.createQueue(MailQueueFactory.SPOOL, MailQueueFactory.prefetchCount(concurrencyLevel));
```

Comment:
I bet you set the prefetch count equal to the concurrency level as rabbitMQ overhead is considered neglectible compared with the mail processing time.

Correct? 

If not, can you provide rationals for this choice?

Reply:
PrefetchCount, for the case of Rabbit will translate to "how many elements can be not-acked-yet at the same time". Any number less than your expected concurrency would make the concurrency less than asked. If you choose a number greater than you concurrency, you will have a less fair distribution of mail dequeues in your cluster.

At first, I was planning something like `concurrency * 1.1` but it doesn't work for small number. Then `concurrency * 1.1 + 1` but I then fell "why should I care for such micro details upfront?" In worst case, we'll have a small pipelining latency. Does it worth the added complexity of having a not-so-clear math formula?

However, I can change my mind you think we should invest in that.

Reply:
No that's perfect, that's a great explanation.

However isn't it a rabbitMQ specific detail somehow being hardcoded in a generic part of the code?

I understand the needs, I just wonder if we have some cleaner ways of linking the implementation to the usage? Maybe with a guice injection?

Reply:
As we use reactive types, it makes sense to let the previous stages of a Publisher/Flux to know how much prefetch we need as it's defined for unitary operations like flatMap, etc.

For now we don't use it on other mailqueues but we probably could if load testing shows specific problems on another implementation we care about.

Reply:
I was thinking: How about using a `ConcurrencyLevel` object in the queue API, then we can have this "prefect = concurrencyLevel" rule in rabbitMQMailQueue, along with a nice comment explaining it? WDYT?

Reply:
I don't get your suggestion. Is it just renaming PrefetchCount type to ConcurrencyLevel? Is it something else?

Reply:
> Is it just renaming PrefetchCount type to ConcurrencyLevel?

Yes.

That allow relocating the actual decision of "the prefetch is the concurrency level" in rabbit backend, and add a comment explaining why.

Sorry I may be picky, but trying to keep things generic enough in APIs seems important to me.

Reply:
A `Flux` doesn't have a concurrency level, you need a `ParallelFlux` to express that.

For now, we only use `Publisher` type and it's closer to `Flux` than `ParallelFlux` so I would not be comfortable to define a concurrency property.

For `Flux` concurrency doesn't exist, only operators have that concept.

However `Flux` does have a `prefetch` concept (how many elements are flowing at a given time).

So i'm picky too but I prefer prefetch. Let me know if you disagree on some of my arguments.

Reply:
We're not speaking of the concurrency level of the resulting flux but that the underlying mailQueue is expected to handle.

Reply:
> We're not speaking of the concurrency level of the resulting flux but that the underlying mailQueue is expected to handle.

The mailqueue doesn't care about concurrency, it's only a consumer concern.

However, how many pending elements you can take from the queue is something the queue has to manage to enable concurrency on the consumer side.

> To be noted, reactor is not even part of the MailQueue API, it's an implementation detail.

Publisher is part of MailQueue API

![Screenshot from 2020-02-28 16-00-00](https://user-images.githubusercontent.com/708318/75559252-0adc4700-5a3b-11ea-95f4-a171bfada138.png)